cpp/kernels/xqa/mla_sm120.cu

```diff
@@ -789,28 +738,67 @@ private:
                 break;
             }
             WarpAcc acc{};
-            using AtomBx2 = Vec<uint32_t, 4>; // one AtomB is 8x32 and AtomBx2 is 16x32
             // wait until it's our turn
             tensorCoreMutex.lock(grpIter);
-            KBarWaiter kBarWaiter{smem, ctaIter, 0};
+            BarWaiter kBarWaiter(smem.kBars, ctaIter * nbKParts + 0);
+            kBarWaiter.testWait();
+            RegQPart regQBuf;
+#if USE_REG_Q
+            static_assert(SharedMemA::regQParts > 0);
+            regQBuf[0] = regQ[0][0];
+#else
+            regQBuf[0] = loadRegQCol(smem.q[0], 0);
+#endif
+            kBarWaiter.wait();
+            RegKPart regKBuf;
+            regKBuf[0] = loadRegKCol(smem.k[kBarWaiter.idxBuf], 0);
+
+            auto shouldTestWait = [](uint32_t idxInstK, uint32_t idxAtomBx2)
+            { return idxInstK == partNbInstK - 1 && idxAtomBx2 == tileNbAtomBx2 - 2; };
+            BarWaiter kBarWaiterNext = kBarWaiter.next();
 #if USE_REG_Q
 #pragma unroll
             for (uint32_t idxPart = 0; idxPart < SharedMemA::regQParts; idxPart++)
             {
-                kBarWaiter.next();
-                kBarWaiter.wait();
 #pragma unroll
-                for (uint32_t idxInstK = 0; idxInstK < exactDiv(partElemsK, qmmaShape.k); idxInstK++)
+                for (uint32_t idxInstK = 0; idxInstK < partNbInstK; idxInstK++)
                 {
-                    Mat16x32Loader const loaderK(kBarWaiter.getK(), 0, idxInstK, rB, cB);
+                    bool const prefetchNextPart = (idxInstK == partNbInstK - 1);
```

Comment:
same here, so your main optimizations are prefetching K tiles to registers as well ?